src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1572,7 +1575,14 @@ static unsigned uct_dc_mlx5_ep_fc_hard_req_progress(void *arg)
      * resend FC_HARD_REQ packet to make sure a peer will resend FC_PURE_GRANT
      * packet in case of failure on the remote FC endpoint */
     kh_foreach_key(&iface->tx.fc_hash, ep_key, {
-        ep     = (uct_dc_mlx5_ep_t*)ep_key;
+        ep = (uct_dc_mlx5_ep_t*)ep_key;
+        uct_dc_mlx5_get_arbiter_params(iface, ep, &waitq, &group, &pool_index);
```

Comment:
@dmitrygx it seems suspicious that this ep was able to allocate a DCI during uct_dc_mlx5_ep_check_fc, but it was also scheduled on the arbiter. so why it didn't allocate the DCI when dispatching the arbiter?
it's similar to a case when we send something on an ep while it's scheduled on the arbiter - we should fail to send because no resources (in this case I would expect that if ep is on the dci arbiter, then no free dci is available anyway)

Reply:
An EP is on arbiter, because it is waiting for FC window to be updated by a peer. But when we send FC_HARD_REQ we don’t check FC resources.

Reply:
but in the issue https://github.com/openucx/ucx/issues/8381 the assertion is from progressing the DCI wait arbiter (uct_dc_mlx5_iface_dci_do_pending_wait) so it means the EP was waiting for DCI allocation.
The issue seems to be the ep was waiting for DCI allocation, and then uct_dc_mlx5_ep_check_fc allocated a DCI for it while it is still on the arbiter.
But that is surprising because I would expect uct_dc_mlx5_ep_check_fc would not be able to allocate a DCI.
Maybe add assertion that when uct_dc_mlx5_ep_check_fc allocates the DCI, the DCI waitq arbiter should be empty, to debug?

Reply:
> The issue seems to be the ep was waiting for DCI allocation, and then uct_dc_mlx5_ep_check_fc allocated a DCI for it while it is still on the arbiter.

right, that's what this PR is fixing

> But that is surprising because I would expect uct_dc_mlx5_ep_check_fc would not be able to allocate a DCI.

why it can't? it doesn't check for FC window resource.
but normal AM/RMA/AMO operations check for FC window resource prior allocating DCI.

> Maybe add assertion that when uct_dc_mlx5_ep_check_fc allocates the DCI, the DCI waitq arbiter should be empty, to debug?

it could be non-empty, because DCI waitq could have other EPs are scheduling for DCIs.